app/cdap/services/react/customHooks/useDebounce.ts

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright © 2021 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+import { useEffect, useState } from 'react';
+
+/**
+ * Custom hook to return a debounced value
+ * @param value - value to return after the delay
+ * @param delay [delay=300] - delay in ms for debounce to wait
+ * @returns debounced value
+ */
+export function useDebounce<T>(value: T, delay = 300): T {
```

Comment:
You've gone back and forth here, so I don't want to belabor the point; consider this optional. But I generally feel we should probably just have a customHooks.ts file that has our general hooks in it at the moment. My take is that if a file has fewer than 200 lines, it's probably too short, but if it's more than 500 lines, it's probably time to break it up.

My thinking here is that a file is effectively a directory of components anyway. Limiting ourselves to one function per file means many files which are very short, like this one. That means more separate files to open sequentially when attempting to take in the codebase. It also means more redundant license and import statements which, if combined into fewer files, would additionally reduce the overall number of human-relevant lines needed to do a full readthrough of the codebase.

That all said, combining two files isn't that big of a deal on its own. Do what you feel like!

Reply:
I'm fine with keeping these as separate. If we keep adding hooks, the file will get large enough to split, and I don't find multiple files in one directory that hard to navigate. Consider also the test files - one set of tests for one hook is about 40 lines (not counting copyright) so that file would grow even quicker than the main file.